File: apps/web/src/services/github.ts
/** GitHub connector + Ask PR approval API. */

import { apiFetch } from "@/lib/api";
import type { ProposedPullRequest } from "@/services/ask";

export type GitHubCapability = "metadata" | "contents" | "pull_requests";

export interface GitHubRepoPreview {
  full_name: string;
  description?: string | null;
  private?: boolean;
  default_branch?: string;
  html_url?: string;
  language?: string | null;
}

export interface GitHubInspectResult {
  login: string;
  name?: string;
  html_url?: string;
  repositories: GitHubRepoPreview[];
}

export interface GitHubConnection {
  connected: boolean;
  provider?: string;
  account_login?: string;
  capabilities?: GitHubCapability[];
  repositories?: string[];
  selected_resource_count?: number;
  setup_status?: string;
}

export interface OpenPullRequestResult {
  status: string;
  owner: string;
  repo: string;
  path: string;
  branch: string;
  base_branch: string;
  pr_number?: number;
  pr_url?: string;
  commit_sha?: string;
}

export async function approveProposedPullRequest(
  proposal: ProposedPullRequest,
): Promise<OpenPullRequestResult> {
  const res = await apiFetch("/github/pull-requests", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      owner: proposal.owner,
      repo: proposal.repo,
      path: proposal.path,
      base_branch: proposal.base_branch,
      branch_name: proposal.branch_name,
      new_content: proposal.new_content,
      file_sha: proposal.file_sha ?? null,
      pr_title: proposal.pr_title,
      pr_body: proposal.pr_body ?? "",
      commit_message: proposal.commit_message || proposal.pr_title,
    }),
  });
  if (!res.ok) {
    let detail = `Could not open pull request (${res.status})`;
    try {
      const body = await res.json();
      detail = (body?.detail as string) || detail;
    } catch {
      /* keep default */
    }
    throw new Error(detail);
  }
  return res.json();
}

async function parseError(res: Response, fallback: string): Promise<string> {
  try {
    const body = await res.json();
    const detail = body?.detail;
    if (typeof detail === "string") return detail;
    return fallback;
  } catch {
    return fallback;
  }
}

export async function inspectGitHubToken(token: string): Promise<GitHubInspectResult> {
  const res = await apiFetch("/github/inspect", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ token }),
  });
  if (!res.ok) throw new Error(await parseError(res, "Could not inspect this GitHub token."));
  return res.json();
}

export async function getGitHubConnection(): Promise<GitHubConnection> {
  const res = await apiFetch("/github/connection");
  if (!res.ok) throw new Error(await parseError(res, "Could not load GitHub connection."));
  return res.json();
}

export async function connectGitHub(input: {
  token: string;
  capabilities: GitHubCapability[];
  repositories: string[];
}): Promise<GitHubConnection> {
  const res = await apiFetch("/github/connection", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
  if (!res.ok) throw new Error(await parseError(res, "Could not save the GitHub connection."));
  return res.json();
}

export async function updateGitHubConnection(input: {
  capabilities: GitHubCapability[];
  repositories: string[];
}): Promise<GitHubConnection> {
  const res = await apiFetch("/github/connection", {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
  if (!res.ok) throw new Error(await parseError(res, "Could not update GitHub permissions."));
  return res.json();
}

export async function disconnectGitHub(): Promise<void> {
  const res = await apiFetch("/github/connection", { method: "DELETE" });
  if (!res.ok) throw new Error(await parseError(res, "Could not disconnect GitHub."));
}

File: apps/api/api/routes/github.py
"""GitHub connector and pull-request approval endpoints."""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel, Field

from github_access import (
    CAP_PULL_REQUESTS,
    connect_github,
    disconnect_github,
    get_github_connection,
    inspect_github_token,
    load_github_access,
    update_github_policy,
)
from github_client import create_pull_request_with_file
from integrations import require_user_context
from rate_limit import ingest_limit, limit

router = APIRouter(prefix="/github", tags=["github"])


class ApprovePullRequestInput(BaseModel):
    """Payload echoed from a proposed Ask PR draft after user approval."""

    owner: str = Field(min_length=1)
    repo: str = Field(min_length=1)
    path: str = Field(min_length=1)
    base_branch: str = Field(default="main")
    branch_name: str = Field(min_length=1)
    new_content: str
    file_sha: str | None = None
    pr_title: str = Field(min_length=1)
    pr_body: str = ""
    commit_message: str = ""


class GitHubTokenInput(BaseModel):
    token: str = Field(min_length=1)


class GitHubConnectInput(BaseModel):
    token: str = Field(min_length=1)
    capabilities: list[str] = Field(min_length=1)
    repositories: list[str] = Field(min_length=1)


class GitHubPolicyInput(BaseModel):
    capabilities: list[str] = Field(min_length=1)
    repositories: list[str] = Field(min_length=1)


@router.post("/inspect")
@limit(ingest_limit)
async def inspect_github(
    request: Request,
    body: GitHubTokenInput,
    _ctx: Annotated[tuple[str, str], Depends(require_user_context)],
) -> dict:
    """Validate a PAT and return repositories it can see. The token is not stored."""

    return await inspect_github_token(body.token)


@router.get("/connection")
async def github_connection(
    ctx: Annotated[tuple[str, str], Depends(require_user_context)],
) -> dict:
    connection = await get_github_connection(*ctx)
    if connection is None:
        return {"connected": False, "provider": "github"}
    return connection


@router.post("/connection")
@limit(ingest_limit)
async def save_github_connection(
    request: Request,
    body: GitHubConnectInput,
    ctx: Annotated[tuple[str, str], Depends(require_user_context)],
) -> dict:
    org_id, user_id = ctx
    return await connect_github(
        org_id=org_id,
        user_id=user_id,
        token=body.token,
        capabilities=body.capabilities,
        repositories=body.repositories,
    )


@router.patch("/connection")
async def patch_github_connection(
    request: GitHubPolicyInput,
    ctx: Annotated[tuple[str, str], Depends(require_user_context)],
) -> dict:
    org_id, user_id = ctx
    return await update_github_policy(
        org_id=org_id,
        user_id=user_id,
        capabilities=request.capabilities,
        repositories=request.repositories,
    )


@router.delete("/connection")
async def delete_github_connection(
    ctx: Annotated[tuple[str, str], Depends(require_user_context)],
) -> dict[str, str]:
    await disconnect_github(*ctx)
    return {"status": "disconnected"}


@router.post("/pull-requests")
async def approve_pull_request(
    request: ApprovePullRequestInput,
    ctx: Annotated[tuple[str, str], Depends(require_user_context)],
) -> dict:
    """Create a branch, commit the approved file change, and open a PR."""

    org_id, user_id = ctx
    access = await load_github_access(org_id, user_id)
    if access is None:
        raise HTTPException(
            status_code=400,
            detail="GitHub is not connected. Add a token in Apps first.",
        )
    denied = access.deny(CAP_PULL_REQUESTS, request.owner, request.repo)
    if denied:
        raise HTTPException(status_code=403, detail=denied)

    return await create_pull_request_with_file(
        token=access.token,
        owner=request.owner,
        repo=request.repo,
        path=request.path,
        new_content=request.new_content,
        base_branch=request.base_branch,
        branch_name=request.branch_name,
        commit_message=request.commit_message or request.pr_title,
        pr_title=request.pr_title,
        pr_body=request.pr_body,
        file_sha=request.file_sha,
    )
